runtime/src/main/cpp/CyclicCollector.cpp

```diff
@@ -0,0 +1,435 @@
+    thiz->gcProcessor();
+    return nullptr;
+  }
+
+  void gcProcessor() {
+     {
+       Locker locker(&lock_);
+       KStdDeque<ObjHeader*> toVisit;
+       KStdUnorderedSet<ObjHeader*> visited;
+       int restartCount = 0;
+       while (!terminateCollector_) {
+         CHECK_CALL(pthread_cond_wait(&cond_, &lock_), "Cannot wait collector condition")
+         if (!shallRunCollector_) continue;
+         atomicSet(&gcRunning_, 1);
+         restartCount = 0;
+        restart:
+         if (restartCount > 10) {
+           COLLECTOR_LOG("wait for some time to avoid GC trashing\n");
+           struct timeval tv;
+           struct timespec ts;
+           long long nsDelta = 1000LL * 1000LL;
+           ts.tv_nsec = (tv.tv_usec * 1000LL + nsDelta) % 1000000000LL;
+           ts.tv_sec = (tv.tv_sec * 1000000000LL + nsDelta) / 1000000000LL;
+           pthread_cond_timedwait(&cond_, &lock_, &ts);
+         }
+         atomicSet(&mutatedAtomics_, 0);
+         visited.clear();
+         toVisit.clear();
+         sideRefCounts_.clear();
+         for (auto* root: rootset_) {
+           // We only care about frozen values here, as only they could become part of shared cycles.
+           if (!root->container()->frozen()) continue;
+           COLLECTOR_LOG("process root %p\n", root);
+           traverseObjectFields(root, [&toVisit, &visited](ObjHeader** location) {
+             ObjHeader* ref = *location;
+             if (ref != nullptr && visited.count(ref) == 0) {
+               toVisit.push_back(ref);
+               COLLECTOR_LOG("adding %p for visiting\n", ref);
+             }
+           });
+           if (atomicGet(&mutatedAtomics_) != 0) {
+             COLLECTOR_LOG("restarted during visiting collect\n")
+             restartCount++;
+             goto restart;
+           }
+           while (toVisit.size() > 0)  {
+             auto* obj = toVisit.front();
+             toVisit.pop_front();
+             COLLECTOR_LOG("visit %p\n", obj);
+             auto* objContainer = obj->container();
+             if (objContainer == nullptr) continue;  // Permanent object.
+             if (!objContainer->frozen()) continue;
+             sideRefCounts_[obj]++;
+             visited.insert(obj);
+             if (atomicGet(&mutatedAtomics_) != 0) {
+               COLLECTOR_LOG("restarted during rootset visit\n")
+               restartCount++;
+               goto restart;
+             }
+             traverseObjectFields(obj, [&toVisit, &visited](ObjHeader** location) {
+               ObjHeader* ref = *location;
+               if (ref != nullptr && visited.count(ref) == 0) {
+                 toVisit.push_back(ref);
+               }
+             });
+           }
+         }
+         // Now find all elements with external references, and mark objects reachable from them as non suitable
+         // for collection by setting their side reference count to -1.
+         toVisit.clear();
+         for (auto it: sideRefCounts_) {
```

Comment:
The entire loop is not properly synchronized with `objContainer->refCount()` mutation.
For example:
```kotlin
import kotlin.native.concurrent.*
  
fun createRef(): AtomicReference<Any?> {
    val atomic1 = AtomicReference<Any?>(null)
    val atomic2 = AtomicReference<Any?>(null)
    atomic1.value = atomic2
    atomic2.value = atomic1
    return atomic1
}

class Holder(var value: AtomicReference<Any?>) {
    fun switch() {
        value = value.value as AtomicReference<Any?>
    }
}

fun createHolder() = Holder(createRef())

fun main() {
    val holder = createHolder()
    kotlin.native.internal.GC.collect()
    kotlin.native.internal.GC.collectCyclic()
    Worker.current.park(1500 * 1000)
    holder.switch()
    kotlin.native.internal.GC.collect()
    Worker.current.park(1500 * 1000)
    Worker.start().executeAfter(0L, {
        kotlin.native.internal.GC.collect()
    }.freeze())
    Worker.current.park(1000)
    println("VALUE: " + holder.value.value)
}  
```
If `sleep(1)` was added to the beginning of the loop's body, then this code would print `VALUE: ... -> null`, indicating that one of the references got nullified.

Reply:
A similar race may still be possible when using local variable instead of `Holder`, since local GC's stack and buffer processing is not actually mutually exclusive with this loop.

Reply:
Kinda reproducer with https://github.com/JetBrains/kotlin-native/commit/765358085afc70a705d83624595e6ca7a1d48e7d 

Doesn't fire in 100% of cases because depends on order of elements in hash table.

```kotlin
import kotlin.native.concurrent.*
import kotlin.native.internal.GC
  
class Node(val value: Any?)

fun createNode1(): Node {
    val ref = AtomicReference<Any?>(null)
    val node2 = Node(ref)
    val node1 = Node(node2)
    ref.value = node1.freeze()

    return node1
}

fun getNode2(): Node {
    val node1 = createNode1()
    GC.collect()

    return node1.value as Node
}   

fun main() {
    val worker = Worker.start()
    
    val node2 = getNode2()
    worker.executeAfter(10 * 1000L, { GC.collectCyclic() }.freeze())

    GC.collect()

    Worker.current.park(50 * 1000L)

    worker.execute(TransferMode.SAFE, {}, {}).result
    
    val ref = node2.value as AtomicReference<Any?>
    check(ref.value != null)
}
```